### PR TITLE
Stage5

### PR DESCRIPTION
**Pull request by KateMoreva**

### CODE REVIEW

**Comment by KateMoreva** [comment]
Результаты профилирования будут добавлены в ближайшее время.

**Comment by KateMoreva** [comment]
### PUT запросы
-t1 -c64 -d5m -R2000 -s ./wrk/put.lua --latency http://localhost:8080

**CPU**


Running 1m test @ http://localhost:8080
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.837ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.78ms    0.89ms  17.07ms   87.16%
    Req/Sec     2.11k   320.81     4.78k    76.37%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.63ms
 75.000%    2.06ms
 90.000%    2.60ms
 99.000%    5.88ms
 99.900%    9.47ms
 99.990%   15.03ms
 99.999%   17.06ms
100.000%   17.09ms
  119711 requests in 1.00m, 7.65MB read
Requests/sec:   1995.17
Transfer/sec:    130.54KB

**Alloc**


Running 1m test @ http://localhost:8080
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.763ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.85ms    1.04ms  21.26ms   92.05%
    Req/Sec     2.11k   395.89     6.11k    77.98%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.68ms
 75.000%    2.07ms
 90.000%    2.66ms
 99.000%    6.86ms
 99.900%   10.85ms
 99.990%   18.80ms
 99.999%   20.22ms
100.000%   21.28ms
  119713 requests in 1.00m, 7.65MB read
Requests/sec:   1995.16
Transfer/sec:    130.54KB

**Lock**


Running 1m test @ http://localhost:8080
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.568ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.67ms    0.87ms  15.22ms   86.03%
    Req/Sec     2.11k   311.10     4.89k    80.10%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.55ms
 75.000%    2.00ms
 90.000%    2.52ms
 99.000%    3.88ms
 99.900%   11.71ms
 99.990%   14.02ms
 99.999%   14.98ms
100.000%   15.23ms
  119710 requests in 1.00m, 7.65MB read
Requests/sec:   1995.14
Transfer/sec:    130.54KB

### GET запросы
-t1 -c64 -d5m -R2000 -s ./wrk/get.lua --latency http://localhost:8080

**CPU**


Running 1m test @ http://localhost:8080
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.738ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.72ms  596.73us  12.81ms   71.15%
    Req/Sec     2.11k   244.12     3.44k    72.00%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.64ms
 75.000%    2.02ms
 90.000%    2.53ms
 99.000%    3.22ms
 99.900%    5.55ms
 99.990%    7.98ms
 99.999%   10.93ms
100.000%   12.81ms
  119710 requests in 1.00m, 11.99MB read
Requests/sec:   1995.16
Transfer/sec:    204.56KB

**Alloc**


Running 1m test @ http://localhost:8080
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.750ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.73ms  582.98us  19.06ms   80.00%
    Req/Sec     2.11k   281.68     4.44k    63.17%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.66ms
 75.000%    2.01ms
 90.000%    2.33ms
 99.000%    3.29ms
 99.900%    6.55ms
 99.990%   14.60ms
 99.999%   17.57ms
100.000%   19.07ms
  119712 requests in 1.00m, 11.99MB read
Requests/sec:   1995.18
Transfer/sec:    204.56KB

**Lock**


Running 1m test @ http://localhost:8080
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.730ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.71ms  599.00us  12.36ms   70.29%
    Req/Sec     2.11k   218.68     3.40k    70.33%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.65ms
 75.000%    2.02ms
 90.000%    2.54ms
 99.000%    3.15ms
 99.900%    5.36ms
 99.990%   10.01ms
 99.999%   11.89ms
100.000%   12.37ms
  119716 requests in 1.00m, 11.99MB read
Requests/sec:   1995.22
Transfer/sec:    204.57KB

Полные результаты представлены [тут ](https://github.com/polis-mail-ru/2020-highload-dht/blob/862fdf19ee88618001ef2724ee350e045dbf30aa/tests/stage5/results.md)

**Comment by KateMoreva** [comment]
PUT

Running 1m test @ http://localhost:8080
  64 threads and 64 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.96ms    6.06ms  50.37ms   92.52%
    Req/Sec   659.64     71.73     1.40k    79.51%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.10ms
 75.000%    2.08ms
 90.000%    4.45ms
 99.000%   31.95ms
 99.900%   40.38ms
 99.990%   47.17ms
 99.999%   49.98ms
100.000%   50.40ms
  2398768 requests in 1.00m, 153.27MB read
Requests/sec:  40033.58
Transfer/sec:      2.56MB
GET

File: src/main/java/ru/mail/polis/service/kate/moreva/Replicas.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * Parses the request to get the needed number of answers (ack) and nodes (from).
+     * */
+    public static Replicas parser(final String replicas) {
+        final List<String> params = Splitter.on('/').splitToList(replicas);
+        if (params.size() != 2) {
+            throw new IllegalArgumentException("Wrong Replica factor: " + replicas);
+        }
+        final int confirmation = Integer.parseInt(params.get(0));
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/Replicas.java:29]: Может `ack` для единообразия?

File: src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    public String primaryFor(@NotNull final ByteBuffer key) {
+        return nodes.get((key.hashCode() & Integer.MAX_VALUE) % nodes.size());
+    }
+
+    @NotNull
+    @Override
+    public Set<String> primaryFor(@NotNull final ByteBuffer key, @NotNull final Replicas replicas) {
+        final Set<String> result = new HashSet<>();
+        int index = key.hashCode() & Integer.MAX_VALUE % nodes.size();
+        while (result.size() < replicas.getFrom()) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java:45]: Учитывая, что мы не валидируем число `from` реплик, то, кажется, мы спокойно можем тут зациклиться навсегда, если `from` больше, чем число нод в нашем кластере.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            log.error("DELETE method failed on /v0/entity for id {}.", key.get(), e);
+            return new Response(Response.INTERNAL_ERROR, Response.EMPTY);
+        }
+    }
+
+    /**
+     * Merges responses for GET request.
+     */
+    public Response mergeResponses(final List<Response> result) {
+        final Map<Response, Integer> responses = new ConcurrentSkipListMap<>(Comparator.comparing(this::getStatus));
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:98]: Почему `ConcurrentSkipListMap`? Тут же последовательный код. И я не понял смысл этого отображения.. У `Response` же не переоопределен даже `equals`/`hashCode`, зачем его как ключ использовать. Поможете разобраться?

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        } catch (IOException e) {
+            handleResponse(session, Response.INTERNAL_ERROR);
+        }
+    }
+
+    /**
+     * Creates new response from @param response and sends it.
+     */
+    public void handleResponse(final HttpSession session, final String response) {
+        try {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:162]: Может весь этот метод выразить через следующий?

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            }
+        }
+        return finalResult;
+    }
+
+    /**
+     * Returns response status.
+     */
+    public String getStatus(final Response response) {
+        return response.getHeaders()[0];
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:120]: Надеждее как-то на числовой статус ориентироваться, а не на порядок хедеров..

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+public class MyRequestHelper {
+    private static final String SERVER_ERROR = "Server error can't send response";
+    private static final String TIMESTAMP = "Timestamp: ";
+    private static final String PROXY_HEADER = "X-Proxy-For:";
+    private static final String NOT_ENOUGH_REPLICAS = "504 Not Enough Replicas";
+    private static final Logger log = LoggerFactory.getLogger(MySimpleHttpServer.class);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:25]: Не тот класс.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                                   final HttpSession session,
+                                   final String response) {
+        try {
+            if (ack < replicas.getAck()) {
+                handleResponse(session, NOT_ENOUGH_REPLICAS);
+            } else {
+                session.sendResponse(new Response(response, Response.EMPTY));
+            }
+        } catch (IOException e) {
+            handleResponse(session, Response.INTERNAL_ERROR);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:154]: Как будто мы потенциально тут можем второй ответ отправлять, если попадем сюда из ветки `else`.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        try {
+            executorService.execute(() -> {
+                if (id.isBlank()) {
+                    log.error("Request with empty id on /v0/entity");
+                    requestHelper.handleResponse(session, Response.BAD_REQUEST);
+                    return;
+                }
+                final boolean isProxy = requestHelper.isProxied(request);
+                final Replicas replicasFactor = isProxy
+                        || replicas == null ? Replicas.quorum(nodeClients.size() + 1) : Replicas.parser(replicas);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:132]: Нужно ли кворум создавать на каждый запрос? Может, закэшируем один объект и будем использовать во всех запросах?

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        requestHelper.sendLoggedResponse(session, requestHelper.mergeResponses(result));
+    }
+
+    private void putMethod(final ByteBuffer key, final Request request, final HttpSession session,
+                           final Replicas replicas, final boolean isProxy) {
+        if (isProxy) {
+            requestHelper.sendLoggedResponse(session, requestHelper.putEntity(key, request));
+            return;
+        }
+        final List<Response> result = replication(requestHelper.putEntity(key, request), request, key, replicas)
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:191]: Я правильно понял, что мы всегда делаем запросы к локальному хранилищу? Даже если, например, `2/3` и мы не являемся праймари нодой для ключа?

**Review by vaddya — CHANGES_REQUESTED** [review]
Есть вопросы по реализации. Прошу прощения за задержку с ревью.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java (hunk 3 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:98]: Мне показалось, что будет удобнее использовать Map, а ConcurrentSkipListMap был выбран больше для однообразия

Comment (vaddya) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:98]: А чем удобнее? Как простой контейнер для списка ответов? Я может не до конца понял, что `maxCount` вообще бывает не равный единице?

**Comment by KateMoreva** [comment]
> Есть вопросы по реализации. Прошу прощения за задержку с ревью.

Спасибо большое за замечания, кажется, все поправила.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java (hunk 3 above)
Comment (vaddya) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java]: Вопрос все еще актуален.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java (hunk 9 above)
Comment (vaddya) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:191]: Еще акутально.

File: src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java (hunk 2 above)
Comment (vaddya) [reply src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java]: Обернуть `if`-ом и вернуть пустой список, кажется, не очень хорошим решением. Это ошбика запроса и она просто проглаывается тем, что вы возвращаете пустой список праймери нод. Нужно проверять входные параметры и бросать исключение.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    public void correctReplication(final int ack,
+                                   final Replicas replicas,
+                                   final HttpSession session,
+                                   final String response) {
+        try {
+            if (ack < replicas.getAck()) {
+                sendLoggedResponse(session, new Response(NOT_ENOUGH_REPLICAS, Response.EMPTY));
+            } else {
+                session.sendResponse(new Response(response, Response.EMPTY));
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:151]: А почему тут тоже не использовать `sendLoggedResponse`?

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java (hunk 9 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:191]: Получается, что да, надо вывести определение праймари до обращения к хранилищу.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java (hunk 10 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java]: Забыла, спасибо.

File: src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java (hunk 2 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java]: Исправила, теперь при запросе на большее число нод, чем реально есть бросается 400.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java (hunk 9 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java]: Исправила, теперь метод вызывается только на праймари ноде.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java (hunk 3 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java]: Реализовала без Map, немного подправив логику ответов. Теперь используется дополнительный заголовок, по которому можно определить статус данных (могила / реальное значение).

File: src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    public String primaryFor(@NotNull final ByteBuffer key) {
+        return nodes.get((key.hashCode() & Integer.MAX_VALUE) % nodes.size());
+    }
+
+    @NotNull
+    @Override
+    public Set<String> primaryFor(@NotNull final ByteBuffer key, @NotNull final Replicas replicas) {
+        final Set<String> result = new HashSet<>();
+        int index = key.hashCode() & Integer.MAX_VALUE % nodes.size();
+            while (result.size() < replicas.getFrom()) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java:45]: Кажется, сделали проверку на стороне вызывающего кода, но добавили странный отступ для цикла :)

Все равно выглядит немножко не безопасно, потому что если (вдруг) появится еще одно место, откуда вызывается этот метод, но без проверки того, что нод в кластере не меньше, чем `replicas.from`, то код здесь зациклиться, это звучит опасно. Лучше проверять инварианты на входные параметры в самом методе.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /**
+     * Subsidiary method to get value.
+     * {@code 200, data} (data is found).
+     * {@code 404} (data is not found).
+     * {@code 500} (internal server error occurred).
+     */
+    public Response getEntity(final ByteBuffer key) {
+        final Cell cell;
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:37]: А зачем отделять объявление переменной и присваивание?

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class MyRequestHelper {
+    private static final String SERVER_ERROR = "Server error can't send response";
+    private static final String TIMESTAMP = "Timestamp: ";
+    private static final String TOMBSTONE = "Tombstone: ";
+    private static final String PROXY_HEADER = "X-Proxy-For:";
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:20]: Тоже может пробел после `:` добавить?

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class MyRequestHelper {
+    private static final String SERVER_ERROR = "Server error can't send response";
+    private static final String TIMESTAMP = "Timestamp: ";
+    private static final String TOMBSTONE = "Tombstone: ";
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:19]: А точно нужен отдельный флаг для этого?
Можно ли считать, что если на `GET` возвращается 404, но с таймстемпом > 0, то это могилка? Хоть и этот подход тоже рабочий, можно оставить.

**Review by vaddya — APPROVED** [review]
По профилированию ОК, по реализации минорные замечения.

Принято.

**Comment by KateMoreva** [comment]
> По профилированию ОК, по реализации минорные замечения.
> 
> Принято.

Спасибо большое, все замечания будут исправлены в реализации 6ого этапа.

**Pull request closed without merging**